src/internal/dependencyGraph.ts

```diff
@@ -0,0 +1,173 @@
+import * as taskTypes from "hardhat/types/builtin-tasks";
+
+import { ResolvedFile, Resolver } from "./resolver";
+import { HardhatError } from "hardhat/internal/core/errors";
+import { ERRORS } from "hardhat/internal/core/errors-list";
+
+export class DependencyGraph implements taskTypes.DependencyGraph {
+  public static async createFromResolvedFiles(
+    resolver: Resolver,
+    resolvedFiles: ResolvedFile[],
+  ): Promise<DependencyGraph> {
+    const graph = new DependencyGraph();
+
+    // TODO refactor this to make the results deterministic
+    await Promise.all(resolvedFiles.map((resolvedFile) => graph._addDependenciesFrom(resolver, resolvedFile)));
+
+    return graph;
+  }
+
+  private _resolvedFiles = new Map<string, ResolvedFile>();
+  private _dependenciesPerFile = new Map<string, Set<ResolvedFile>>();
+
+  // map absolute paths to source names
+  private readonly _visitedFiles = new Map<string, string>();
+
+  private constructor() {}
+
+  public getResolvedFiles(): ResolvedFile[] {
+    return Array.from(this._resolvedFiles.values());
+  }
+
+  public has(file: ResolvedFile): boolean {
+    return this._resolvedFiles.has(file.sourceName);
+  }
+
+  public isEmpty(): boolean {
+    return this._resolvedFiles.size === 0;
+  }
+
+  public entries(): Array<[ResolvedFile, Set<ResolvedFile>]> {
+    return Array.from(this._dependenciesPerFile.entries()).map(([key, value]) => [
+      this._resolvedFiles.get(key)!,
+      value,
+    ]);
+  }
+
+  public getDependencies(file: ResolvedFile): ResolvedFile[] {
+    const dependencies = this._dependenciesPerFile.get(file.sourceName) ?? new Set();
+
+    return [...dependencies];
+  }
+
+  public getTransitiveDependencies(file: ResolvedFile): taskTypes.TransitiveDependency[] {
+    const visited = new Set<ResolvedFile>();
+
+    const transitiveDependencies = this._getTransitiveDependencies(file, visited, []);
+
+    return [...transitiveDependencies];
+  }
+
+  public getConnectedComponents(): DependencyGraph[] {
```

Comment:
This function is not used. Do we need it?

Reply:
No, we don't need this functionality in the current version, so I removed this function